File: App/src/Adaptive.ReactiveTrader.Client/UI/Connectivity/ConnectivityStatusViewModel.cs
using System;
using System.Reactive.Linq;
using Adaptive.ReactiveTrader.Client.Domain;
using Adaptive.ReactiveTrader.Client.Domain.Transport;
using Adaptive.ReactiveTrader.Client.Instrumentation;
using Adaptive.ReactiveTrader.Shared.UI;
using log4net;

namespace Adaptive.ReactiveTrader.Client.UI.Connectivity
{
    class ConnectivityStatusViewModel : ViewModelBase, IConnectivityStatusViewModel
    {
        private readonly IPriceLatencyRecorder _priceLatencyRecorder;
        private static readonly ILog Log = LogManager.GetLogger(typeof(ConnectivityStatusViewModel));

        public ConnectivityStatusViewModel(IReactiveTrader reactiveTrader, IPriceLatencyRecorder priceLatencyRecorder)
        {
            _priceLatencyRecorder = priceLatencyRecorder;
            reactiveTrader.ConnectionStatus
                .ObserveOnDispatcher()
                .Subscribe(
                OnStatusChange,
                ex => Log.Error("An error occured within the connection status stream.", ex));

            Observable
                .Timer(TimeSpan.FromSeconds(1))
                .Repeat()
                .ObserveOnDispatcher()
                .Subscribe(OnTimerTick);
        }

        private void OnTimerTick(long _)
        {
            var current = _priceLatencyRecorder.GetCurrentAndReset();
            UiLatency = (int)current.Item1.TotalMilliseconds;
            Throughput = current.Item2;
        }

        private void OnStatusChange(ConnectionInfo connectionInfo)
        {
            switch (connectionInfo.ConnectionStatus)
            {
                case ConnectionStatus.Uninitialized:
                case ConnectionStatus.Connecting:
                    Status = string.Format("Connecting to {0} ...", connectionInfo.Server);
                    break;
                case ConnectionStatus.Reconnected:
                case ConnectionStatus.Connected:
                    Status = string.Format("Connected to {0}", connectionInfo.Server);
                    break;
                case ConnectionStatus.ConnectionSlow:
                    Status = string.Format("Slow connection detected with {0}", connectionInfo.Server);
                    break;
                case ConnectionStatus.Reconnecting:
                    Status = string.Format("Reconnecting to {0} ...", connectionInfo.Server);
                    break;
                case ConnectionStatus.Closed:
                    Status = string.Format("Disconnected from {0}", connectionInfo.Server);
                    break;
                default:
                    throw new ArgumentOutOfRangeException();
            }
        }

        public string Status { get; private set; }
        public long UiLatency { get; private set; }
        public long Throughput { get; private set; }
    }
}

File: src/Adaptive.ReactiveTrader.Client.JS/UI/Connectivity/ConnectivityStatusViewModel.ts
class ConnectivityStatusViewModel implements IConnectivityStatusViewModel {
    private _priceLatencyRecorder: IPriceLatencyRecorder;

    status: KnockoutObservable<string>;
    uiLatency: KnockoutObservable<number>;
    throughput: KnockoutObservable<number>;
    disconnected: KnockoutObservable<boolean>;

    constructor(reactiveTrader: IReactiveTrader, priceLatencyRecorder: IPriceLatencyRecorder) {
        this._priceLatencyRecorder = priceLatencyRecorder;
        reactiveTrader.connectionStatusStream
            .subscribe(
                status=> this.onStatusChanged(status),
                ex=> console.error("An error occured within the connection status stream", ex));

        Rx.Observable
            .timer(1000, Rx.Scheduler.timeout)
            .repeat()
            .subscribe(_=> this.onTimerTick());

        this.status = ko.observable("Disconnected.");
        this.uiLatency = ko.observable(0);
        this.throughput = ko.observable(0);
        this.disconnected = ko.observable(false);
    }

    private onTimerTick() {
        var current = this._priceLatencyRecorder.getMaxLatencyAndReset();
        if (current == null || current.priceWithMaxLatency == null) return;

        this.uiLatency(current.priceWithMaxLatency.uiProcessingTimeMs);
        this.throughput(current.count);
    }

    private onStatusChanged(connectionInfo: ConnectionInfo) {
        switch(connectionInfo.connectionStatus) {
            case ConnectionStatus.Uninitialized:
            case ConnectionStatus.Connecting:
                this.status("Connecting to " + connectionInfo.server + "...");
                this.disconnected(true);
                break;
            case ConnectionStatus.Reconnected:
            case ConnectionStatus.Connected:
                this.status("Connected to " + connectionInfo.server);
                this.disconnected(false);
                break;
            case ConnectionStatus.ConnectionSlow:
                this.status("Slow connection detected with " + connectionInfo.server);
                this.disconnected(false);
                break;
            case ConnectionStatus.Reconnecting:
                this.status("Reconnecting to " + connectionInfo.server + "...");
                this.disconnected(true);
                break;
            case ConnectionStatus.Closed:
                this.status("Disconnected from " + connectionInfo.server);
                this.disconnected(true);
                break;

        }
    }
} 